Refactor CogManager and extension loading

Tobotimus opened the pull request:
This kicked off with me realising our use of [`Loader.load_module()`](https://docs.python.org/3/library/importlib.html#importlib.machinery.SourceFileLoader.load_module) as a hacky way to import modules from random locations is deprecated. Looking closer, I realised there were a couple of smelly things here which could be cleaned up. Here's what's changed:
- Paths in CogManager are now inserted into `redbot.ext_cogs.__path__`. This means to import any external cog (from install path, cogmanager path, whatever), the import path would be `redbot.ext_cogs.cog_to_import`. It makes the rest of the logic really straightforward.
- Importing cog modules is now done with `importlib.import_module` - this is a high-level function which allows us to delegate all of the nitty gritty import logic to Python itself, and it should be much more reliable. We're also no longer dealing with `ModuleSpec` objects, we just go straight from cog name to module.
- We're now using `importlib.reload` to reload modules like a good boy. This is done from the bottom-up, twice, to make sure *most* internal imports in the package are updated. Some old imports could hang around if people do lazy relative importing at the module level.
- `Red.load_extension` can take either a module, or a straight name. This means loading a package is as simple as `Red.load_extension(package_name)`, no more needing to know what method of `CogManager` to use.
- Some internal documentation was added in a module docstring of `redbot.core.cog_manager`.

Tobotimus committed:
Refactor CogManager and extension loading

This kicked off with me realising our use of `Loader.load_module()` as a hacky way to import modules from random locations is deprecated. Looking closer, I realised there were a couple of smelly things here which could be cleaned up. Here's what's changed:
- Paths in CogManager are now inserted into `redbot.cogs.__path__`. This means to import any cog, including an external cog, the import path would be `redbot.cogs.cog_to_import`. It makes the rest of the logic really straightforward. The only side effect is, to differentiate core cogs from external cogs, we now need to look at a module's `__file__` rather than its `__module__` or `__package__`.
- Importing cog modules is now done with `importlib.import_module` - this is a high-level function which allows us to delegate all of the nitty gritty import logic to Python itself, and it should be much more reliable. We're also no longer dealing with `ModuleSpec` objects, we just go straight from cog name to module.
- Since lazy importing is used for cog modules, and because big packages can take quite a long time to import, whenever we're importing a cog module which is not already in `sys.modules`, we do it in a ThreadPoolExecutor to make it non-blocking.
- We're now using `importlib.reload` to reload modules like a good boy. This is done from the bottom-up, twice, to make sure all internal imports in the package are updated.
- Packages are no longer deleted from `sys.modules` in `Red.unload_extension()`. This was really smelly and, in my opinion, unnecessary. It would have resulted in a miniscule memory saving at best, although I'm pretty sure even after doing this the module would still remain in memory, as Python keeps a reference to imported modules in C code as well.
- `Red.load_extension` can take either a module, or a straight name. This means loading a package is as simple as `Red.load_extension(package_name)`, no more needing to know what method of `CogManager` to use.
- Some internal documentation was added in a module docstring of `redbot.core.cog_manager`.

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Do reloading in a ThreadPoolExecutor too

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Fall back to importing in main thread when executor fails

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Merge branch 'V3/develop' into V3/cogmgr_refactor

Tobotimus committed:
Merge branch 'V3/develop' into V3/cogmgr_refactor

# Conflicts:
#	redbot/__main__.py

calebj commented:
I haven't noticed any issues so far in using this, so it seems good. However, I would like to see some more widespread ~~battlefield~~ testing with a variety of third party cogs and external modules before giving my :+1:.

Tobotimus committed:
Merge branch 'V3/develop' into V3/cogmgr_refactor

Tobotimus committed:
Don't use redbot.cogs.__path__ for 3rd party cogs

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Use str Enum for InstallableType

This just makes equality checks play nicer. There was some weird shit happening when we reloaded downloader, causing things like `InstallableType.COG != InstallableType.COG`, because one instance would be from before the reload.

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Merge remote-tracking branch 'origin/V3/cogmgr_refactor' into V3/cogmgr_refactor

Tobotimus committed:
Don't load modules in a secondary thread

Was causing some weird SSL errors and it's probably not worth it anyway.

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Merge branch 'V3/develop' into V3/cogmgr_refactor

Tobotimus committed:
Merge branch 'V3/develop' into V3/cogmgr_refactor

# Conflicts:
#	redbot/core/core_commands.py

Tobotimus committed:
Fix up errors from merge

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus commented:
This is ready for review again and will be targetting 3.2

Tobotimus committed:
Undo accidental changes

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Remove unnecessary CogManager._loop attribute

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Merge branch 'V3/develop' into V3/cogmgr_refactor

Tobotimus committed:
Add towncrier entries

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Reorganise towncrier entries

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Commit:
Merge branch 'V3/develop' into V3/cogmgr_refactor

mikeshardmind commented:
I still have a few concerns with regard to using `importlib.reload` in this manner, but I also haven't seen this break any cogs which I use in any way which wouldn't already be broken with the current reload strategy. I'd like to do another pass over this later before merging it, but it should make the reload behavior of cogs more consistent from an end user standpoint at least.

Commit:
Merge branch 'V3/develop' into V3/cogmgr_refactor

mikeshardmind commented on redbot/core/bot.py:
```diff
@@ -590,28 +591,35 @@ def list_packages():
-        if name in self.extensions:
-            raise errors.PackageAlreadyLoaded(spec)
+        if isinstance(module, str):
+            module: types.ModuleType = self.cog_mgr.load_cog_module(module)
```
```suggestion
            module: types.ModuleType = self._cog_mgr.load_cog_module(module)
```

mikeshardmind commented on redbot/core/cog_manager.py:
```diff
@@ -392,12 +377,7 @@ class CogManagerUI(commands.Cog):
-            await ctx.send(_("Invalid 'to' index."))
-            return
-
+        await ctx.bot.cog_mgr.reorder_path(to_move, to)
```
```suggestion
        await ctx.bot._cog_mgr.reorder_path(to_move, to)
```

mikeshardmind commented on redbot/core/core_commands.py:
```diff
@@ -75,35 +77,35 @@ def __init__(self, bot: "Red"):
-                    cogspecs.append((spec, name))
-                else:
-                    notfound_packages.append(name)
+                module = bot.cog_mgr.load_cog_module(name)
```
```suggestion
                module = bot._cog_mgr.load_cog_module(name)
```

mikeshardmind commented on redbot/core/core_commands.py:
```diff
@@ -75,35 +77,35 @@ def __init__(self, bot: "Red"):
             try:
-                self._cleanup_and_refresh_modules(spec.name)
-                await bot.load_extension(spec)
+                module = bot.cog_mgr.reload(module)
```
```suggestion
                module = bot._cog_mgr.reload(module)
```

mikeshardmind commented on redbot/core/core_commands.py:
```diff
@@ -2198,14 +2173,9 @@ class Core(commands.Cog, CoreLogic):
-
-        await self.bot.load_extension(spec)
+        module = self.bot.cog_mgr.load_cog_module(cog_name)
+        module = self.bot.cog_mgr.reload(module)
```
```suggestion
        module = self.bot._cog_mgr.reload(module)
```

mikeshardmind commented on redbot/core/core_commands.py:
```diff
@@ -2198,14 +2173,9 @@ class Core(commands.Cog, CoreLogic):
-        self._cleanup_and_refresh_modules(spec.name)
-
-        await self.bot.load_extension(spec)
+        module = self.bot.cog_mgr.load_cog_module(cog_name)
```
```suggestion
        module = self.bot._cog_mgr.load_cog_module(cog_name)
```

Drapersniper commented:
@Tobotimus is this something you are still actively working on or have you dropped this PR. Unsure whether you are currently working on this locally and just haven't pushed any commits.

Drapersniper commented:
Closing due to innactivity, Will convert into an issue so that other users can pick this up.
